internal/util/trace/span.go

```
@@ -0,0 +1,68 @@
+package trace
+
+import (
+	"encoding/json"
+
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/trace"
+)
+
+type Span struct {
+	trace.Span
+}
+
+func (s *Span) RecordInt64Pairs(keys []string, values []int64) {
+	if len(keys) != len(values) {
+		return
+	}
+	attrs := make([]attribute.KeyValue, 0, len(keys))
+	for i := range keys {
+		attrs = append(attrs, attribute.Int64(keys[i], values[i]))
+	}
+	s.SetAttributes(attrs...)
+}
+
+func (s *Span) RecordInt(key string, val int) {
+	s.SetAttributes(attribute.Int(key, val))
+}
+
+func (s *Span) RecordInt64(key string, val int64) {
+	s.SetAttributes(attribute.Int64(key, val))
+}
+
+func (s *Span) RecordInts(key string, val []int) {
+	s.SetAttributes(attribute.IntSlice(key, val))
+}
+
+func (s *Span) RecordInt64s(key string, val []int64) {
+	s.SetAttributes(attribute.Int64Slice(key, val))
+}
+
+func (s *Span) RecordString(key, val string) {
+	s.SetAttributes(attribute.String(key, val))
+}
+
+func (s *Span) RecordStrings(key string, val []string) {
+	s.SetAttributes(attribute.StringSlice(key, val))
+}
+
+func (s *Span) RecordStringPairs(keys []string, vals []string) {
+	attrs := make([]attribute.KeyValue, 0, len(keys))
+	for i := range keys {
+		attrs = append(attrs, attribute.String(keys[i], vals[i]))
+	}
+	s.SetAttributes(attrs...)
+}
+
+func (s *Span) RecordBool(key string, val bool) {
+	s.SetAttributes(attribute.Bool(key, val))
+}
+
+func (s *Span) RecordAny(key string, val any) {
+	payload, err := json.Marshal(val)
+	if err != nil {
+		s.SetAttributes(attribute.String(key, "{}"))
+	} else {
```

Need some method to detect serialization failure other than swallow this error.